File: src/com/colacelli/ircbot/plugins/access/IRCBotAccess.java
package com.colacelli.ircbot.plugins.access;

import com.colacelli.ircbot.IRCBot;
import com.colacelli.ircbot.listeners.OnChannelCommandListener;
import com.colacelli.irclib.actors.User;
import com.colacelli.irclib.connection.Connection;
import com.colacelli.irclib.connection.Rawable;
import com.colacelli.irclib.connection.listeners.OnRawCodeListener;
import com.colacelli.irclib.messages.ChannelMessage;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.HashMap;
import java.util.Properties;

public class IRCBotAccess {
    public static final int SUPER_ADMIN_LEVEL = 10;
    public static final int ADMIN_LEVEL = 5;
    public static final int OPERATOR_LEVEL = 1;

    private static final String PROPERTIES_FILE = "access.properties";

    private static IRCBotAccess instance;
    private Properties properties;


    private IRCBotAccess() {
        properties = new Properties();
    }

    public static IRCBotAccess getInstance() {
        if (instance == null) {
            instance = new IRCBotAccess();
        }

        return instance;
    }

    public void addListener(IRCBot bot, int access, OnChannelCommandListener listener) {
        bot.addListener(new OnChannelCommandListener() {
            @Override
            public String channelCommand() {
                return listener.channelCommand();
            }

            @Override
            public void onChannelCommand(Connection connection, ChannelMessage message, String command, String... args) {
                if (IRCBotAccess.getInstance().getLevel(message.getSender()) >= access) {
                    // Check by whois if nick is identified
                    OnRawCodeListener rawCodeListener = new OnRawCodeListener() {
                        @Override
                        public int rawCode() {
                            return Rawable.RawCode.WHOIS_IDENTIFIED_NICK.getCode();
                        }

                        @Override
                        public void onRawCode(Connection connection, String message1, int rawCode, String... args1) {
                            // Same nickname?
                            if (args1[3].equals(message.getSender().getNick())) {
                                listener.onChannelCommand(connection, message, command, args);
                            }
                        }
                    };

                    bot.addListener(rawCodeListener);

                    // Whois end?
                    bot.addListener(new OnRawCodeListener() {
                        @Override
                        public int rawCode() {
                            return Rawable.RawCode.WHOIS_END.getCode();
                        }

                        @Override
                        public void onRawCode(Connection connection1, String message1, int rawCode, String... args1) {
                            // Same nickname?
                            if (args1[3].equals(message.getSender().getNick())) {
                                // Remove whois listeners
                                bot.removeListener(rawCodeListener);
                                bot.removeListener(this);
                            }
                        }
                    });

                    connection.whois(message.getSender());
                }
            }
        });
    }

    public void removeListener(IRCBot bot, String command) {
        bot.removeListener(command);
    }

    public void setLevel(String nick, int level) {
        loadProperties();

        nick = nick.toUpperCase();

        if (level > 0) {
            properties.setProperty(nick, String.valueOf(level));
        } else {
            properties.setProperty(nick, "");
        }

        saveProperties();
    }

    public int getLevel(User user) {
        return getLevel(user.getNick());
    }

    public int getLevel(String nick) {
        loadProperties();

        int level;
        try {
            level = Integer.parseInt(properties.getProperty(nick.toUpperCase()));
        } catch (NumberFormatException e) {
            level = 0;
        }

        return level;
    }

    public HashMap<String, Integer> getAccesses() {
        loadProperties();

        HashMap<String, Integer> accesses = new HashMap<>();
        properties.forEach((key, value) -> {
            try {
                accesses.put(String.valueOf(key), Integer.valueOf(String.valueOf(value)));
            } catch (NumberFormatException e) {
                // Invalid level
            }
        });

        return accesses;
    }

    private void loadProperties() {
        try {
            FileInputStream fileInputStream = new FileInputStream(PROPERTIES_FILE);
            properties.load(fileInputStream);
        } catch (IOException e) {
            // Properties file not found
            properties = new Properties();
            saveProperties();
        }
    }

    private void saveProperties() {
        OutputStream outputStream = null;

        try {
            outputStream = new FileOutputStream(PROPERTIES_FILE);
            properties.store(outputStream, null);
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                outputStream.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}

File: src/com/colacelli/ircbot/plugins/access/IRCBotAccess.kt
package com.colacelli.ircbot.plugins.access

import com.colacelli.ircbot.IRCBot
import com.colacelli.ircbot.listeners.OnChannelCommandListener
import com.colacelli.ircbot.PropertiesPlugin
import com.colacelli.irclib.actors.User
import com.colacelli.irclib.connection.Connection
import com.colacelli.irclib.connection.Rawable
import com.colacelli.irclib.connection.listeners.OnRawCodeListener
import com.colacelli.irclib.messages.ChannelMessage
import com.colacelli.irclib.messages.PrivateNoticeMessage
import java.util.*
import kotlin.collections.HashMap

class IRCBotAccess : PropertiesPlugin {
    var properties = Properties()

    enum class Level(val level : Int) {
        ROOT(3),
        ADMIN(2),
        OPERATOR(1),
        USER(0)
    }

    private object Singleton {
        val instance = IRCBotAccess()
    }

    companion object {
        const val PROPERTIES_FILE = "access.properties"

        val instance by lazy {
            Singleton.instance
        }
    }

    fun addListener(bot : IRCBot, level : Level, listener: OnChannelCommandListener) {
        bot.addListener(object : OnChannelCommandListener {
            override fun channelCommand(): String {
                return listener.channelCommand()
            }

            override fun onChannelCommand(connection: Connection, message: ChannelMessage, command: String, args: Array<String>) {
                if (get(message.sender!!).level >= level.level) {
                    // Identified by nickserv?
                    val identifiedListener = object : OnRawCodeListener {
                        override fun rawCode(): Int {
                            return Rawable.RawCode.WHOIS_IDENTIFIED_NICK.code
                        }

                        override fun onRawCode(connection: Connection, message1: String, rawCode: Int, args1: List<String>) {
                            if (args1[3] == message.sender?.nick) {
                                listener.onChannelCommand(connection, message, command, args)
                            }
                        }

                    }
                    bot.addListener(identifiedListener)

                    // End of whois?
                    bot.addListener(object : OnRawCodeListener {
                        override fun rawCode(): Int {
                            return Rawable.RawCode.WHOIS_END.code
                        }

                        override fun onRawCode(connection: Connection, message1: String, rawCode: Int, args1: List<String>) {
                            if (args1[3] == message.sender?.nick) {
                                bot.removeListener(identifiedListener)
                                bot.removeListener(this)
                            }
                        }

                    })

                    connection.whois(message.sender!!)
                } else {
                    val response = PrivateNoticeMessage("You don't have access to that command!", connection.user, message.sender)
                    connection.send(response)
                }
            }
        })
    }

    fun removeListener(bot: IRCBot, command: String) {
        bot.removeListener(command)
    }

    fun get(user : User) : Level {
        properties = loadProperties(PROPERTIES_FILE)

        return Level.valueOf(properties.getProperty(user.nick.toLowerCase()))
    }

    fun add(nick:String, level:Level) {
        properties = loadProperties(PROPERTIES_FILE)
        properties.setProperty(nick.toLowerCase(), level.toString())
        saveProperties(PROPERTIES_FILE, properties)
    }

    fun del(nick:String) {
        properties = loadProperties(PROPERTIES_FILE)
        properties.remove(nick.toLowerCase())
        saveProperties(PROPERTIES_FILE, properties)
    }

    fun list() : SortedMap<String, Level> {
        properties = loadProperties(PROPERTIES_FILE)

        val accesses = HashMap<String, Level>()
        properties.forEach {
            accesses[it.key.toString()] = Level.valueOf(it.value.toString())
        }

        return accesses.toSortedMap()
    }
}
